42/README.md

```diff
@@ -0,0 +1,71 @@
+```
+shortname: [42/HARD-FORKS]
+name: Handling Hard Forks
+type: information
+Status: raw
+editor: Vanshdeep Singh <vanshdeep@bigchaindb.com>
+```
+
+## Description
+
+Currently, there is no active strategy to address breaking changes in BigchainDB 2.0 i.e. there is no general approach to migration such that breaking changes in transaction specs can be handled. This document is an attempt in quantifying scenarios wherein a breaking might occur and propose a generalized strategy to handle these situations.
+
+## Hard fork scenarios
+1. Change in any of the transaction specs
+    - Change in signature algorithm
+    - Change in hashing functions
+    - Change in schema
+2. Change in storage schema
+
+## Handling Hard Forks
+
+### Type 1 scenarios
+In case of change to spec it might not be possible to port all existing validated transaction to the new format. In such cases one needs to make a distinction between transactions which have already been validated and stored in the Blockchain from new incoming transactions. Furthermore, a suitable trigger is needed so that the network can collectively jump to the new logic else there could be situations wherein the network might not be able to commit any new blocks because of honest nodes running different validation logic. So, it is apparent that a mechanism for transitioning to this new validation logic is need. Below is one such procedure to achieve the same,
+
+- In order to decide when the validation logic should be upgraded to its newer version a TEP is initiated which defines the future block `height` at which to upgrade and the corresponding `BigchainDB` class to use.
```

Comment:
Just some thoughts: it would be cool to be able to fork off of a height from the past as well. For example, as soon as an election with the height from the past is concluded, the app abandons the corresponding part of the chain and starts to build up from this height again. But I am afraid it can be done only in sync with Tendermint so there should be a similar feature there, right?

Reply:
So, a theoretical Tendermint feature may allow to avoid replaying this tail. But you are right - the initial step towards the solution can be done with what we've got already!

> allowing for such a behaviour would be an anti-pattern

If there is a good reason for such a fork, people will make it anyway but they would need to orchestrate it themselves. Why do you think envisaging the tooling for that is an antipattern?

Reply:
The idea you are a suggestion here stems from the desire to delete data from a blockchain. Since blockchains are _viewed as immutable datastores_, such a feature would seem anti-pattern. An argument in favour of such a feature comes about when we tweak the definition of blockchain a little i.e. software to reach consensus in presence of byzantine actors (which is what they actually are). So, if the network agrees to undo `m` blocks it should be able to do so.

All in all what I wish to state is that your suggestion is aiming to solve a subset or a special case of a general problem. If the users ask for a solution for this specific use case then we can implement such a feature but until then we can let it be.